ntp-daemon/src/main.rs

```diff
@@ -23,5 +27,66 @@ async fn main() -> Result<(), Box<dyn Error>> {
     // this away when not using sentry.
     let _guard = finish_tracing_init(&mut config, has_log_override)?;
 
-    ntp_daemon::spawn(&config.system, &config.peers).await
+    let peers_reader = Arc::new(tokio::sync::RwLock::new(Vec::new()));
+    let peers_writer = peers_reader.clone();
+
+    std::fs::remove_dir_all("/run/ntpd-rs")?;
+    std::fs::create_dir("/run/ntpd-rs")?;
+
+    let peers_listener = UnixListener::bind("/run/ntpd-rs/observe")?;
+
+    // this binary needs to run as root to be able to adjust the system clock
+    // directories inherit root permissions, but the client should not need elevated
+    // permissions to read from the socket. So we explicitly set the permissions
+    let permissions: std::fs::Permissions = PermissionsExt::from_mode(0o777);
+
+    std::fs::set_permissions("/run/ntpd-rs", permissions.clone())?;
+    std::fs::set_permissions("/run/ntpd-rs/observe", permissions)?;
+
+    let handle = tokio::spawn(async move {
+        ntp_daemon::spawn(&config.system, &config.peers, peers_writer).await
+    });
+
+    // to prevent the handle being consumed in the first loop iteration
+    tokio::pin!(handle);
+
+    loop {
+        tokio::select! {
+            done = (&mut handle) => {
+                return Ok(done??);
+            }
+            accept = peers_listener.accept() => {
+                let (stream, _addr) = accept?;
+
+                let buffer = {
+                    let state = peers_reader.read().await;
+
+                    serde_json::to_vec(state.deref()).unwrap()
+                };
+
+                write_to_unix_socket(stream, &buffer).await?;
+            }
+        }
```

Comment:
I think this may become a little bit nicer if we spawn the observability handler as its own task. At that point we could wait for either to finish, or both with simpler constructs like join.

Reply:
I think we want `either` behavior here. Certainly, when the system task is done, the application should shut down. Any of the other observing tasks should only exit when they hit an error, which I think should also just bring the app down (this most likely happens on startup, when some path is not right or some configuration setting is incorrect/missing)